File: ext/carray_undef.c
/* ---------------------------------------------------------------------------

  UNDEF sentinel value: the single instance of UndefClass exported as
  the top-level constant `UNDEF`.  Used by mask-bearing CArray APIs
  to mean "masked / no value here" without choosing a numeric sentinel
  (NaN / -1 / etc.) that might collide with valid data.

  Identity semantics: every reference to UNDEF is the same Ruby object;
  C code compares with raw pointer equality (`rval == CA_UNDEF`).  The
  GC pin in Init_carray_undef guarantees the pointer stays valid even
  under Ruby 3+ compacting GC.

  Coercion asymmetry: to_s / inspect return the literal string "UNDEF"
  (display works), but to_f / to_i / to_int raise TypeError (numeric
  coercion is rejected so accidental arithmetic on UNDEF surfaces as
  an error rather than silently producing 0 / NaN).

---------------------------------------------------------------------------- */

#include "ruby.h"

static VALUE rb_cUNDEF;

VALUE CA_UNDEF;

static VALUE rb_ud_inspect (VALUE self)
{
  return rb_str_new2("UNDEF");
}

static VALUE rb_ud_to_s (VALUE self)
{
  return rb_str_new2("UNDEF");
}

static VALUE rb_ud_to_f (VALUE self)
{
  rb_raise(rb_eTypeError, "can't coerce UNDEF into Float");
}

static VALUE rb_ud_to_i (VALUE self)
{
  rb_raise(rb_eTypeError, "can't coerce UNDEF into Integer");
}

static VALUE rb_ud_equal (VALUE self, VALUE other)
{
  return ( self == other ) ? Qtrue : Qfalse;
}

void
Init_carray_undef (void)
{
  rb_cUNDEF = rb_define_class("UndefClass", rb_cObject);
  rb_define_method(rb_cUNDEF, "inspect", rb_ud_inspect, 0);
  rb_define_method(rb_cUNDEF, "to_s",    rb_ud_to_s,    0);
  rb_define_method(rb_cUNDEF, "to_f",    rb_ud_to_f,    0);
  rb_define_method(rb_cUNDEF, "to_i",    rb_ud_to_i,    0);
  rb_define_method(rb_cUNDEF, "to_int",  rb_ud_to_i,    0);
  rb_define_method(rb_cUNDEF, "==",      rb_ud_equal,   1);

  /* Singleton: instantiate once, then undef `new` so no second instance
     can be created.  The identity invariant relies on this. */
  CA_UNDEF = rb_funcall(rb_cUNDEF, rb_intern("new"), 0);
  rb_undef_method(CLASS_OF(rb_cUNDEF), "new");
  rb_const_set(rb_cObject, rb_intern("UNDEF"), CA_UNDEF);

  /* CAREFUL: CA_UNDEF is dereferenced via raw pointer compare
     (`rval == CA_UNDEF`) throughout the C code, so the underlying
     object must not be relocated.  Constant registration above
     keeps the object reachable but does not prevent compacting GC
     (Ruby 2.7+) from moving it to a new address; the pin here
     disables compaction for this single sentinel. */
  rb_gc_register_mark_object(CA_UNDEF);
}

File: yard-stubs/carray_undef.rb
# DOCUMENTATION ONLY — DO NOT REQUIRE.
# Stubs for the UNDEF sentinel and UndefClass defined in
# ext/carray_undef.c. See yard-stubs/README.md and yard-stubs/STYLE.md.

# Singleton class of the {UNDEF} sentinel value.  Not instantiable
# (`UndefClass.new` is undef'd after the single instance is created
# in `Init_carray_undef`).
#
# The class is private API in practice: read and write the sentinel
# through the top-level `UNDEF` constant rather than touching the
# class directly.
class UndefClass
  # @overload inspect
  #   Returns the literal string `"UNDEF"`.
  #   @return [String]
  def inspect; end

  # @overload to_s
  #   Returns the literal string `"UNDEF"`.  Display works (so
  #   `puts UNDEF` shows `"UNDEF"`); numeric coercion via {#to_f}
  #   / {#to_i} raises.
  #   @return [String]
  def to_s; end

  # @overload to_f
  #   Raises `TypeError`.  Numeric coercion of UNDEF is rejected so
  #   that accidental arithmetic on a masked sentinel surfaces as an
  #   error rather than silently producing `0.0` / `NaN`.
  #   @raise [TypeError]
  def to_f; end

  # @overload to_i
  #   Raises `TypeError`.  Same rationale as {#to_f}.
  #   @raise [TypeError]
  def to_i; end

  # @overload to_int
  #   Alias of {#to_i}.  Raises `TypeError`.
  #   @raise [TypeError]
  def to_int; end

  # @overload ==(other)
  #   Returns `true` only when `other` is the same UNDEF singleton
  #   (identity comparison).  No coercion or value equality.
  #   @param other [Object]
  #   @return [Boolean]
  def ==(other); end
end

# Top-level sentinel used by mask-bearing CArray APIs to mean
# "masked / no value here" without choosing a numeric sentinel
# (NaN, `-1`, etc.) that might collide with valid data.
#
# Identity semantics: every reference to `UNDEF` is the same Ruby
# object, pinned against the compacting GC so the C extension can
# safely use raw pointer comparison to detect it.
#
# @example mask a cell during construction
#   a = CArray.int32(3) { |i| (i == 1) ? UNDEF : i * 10 }
#   a.to_a  # => [0, nil, 20]    # UNDEF surfaces as nil after to_a
#
# @example check whether an element is masked
#   a[1] == UNDEF  # => true
UNDEF = UndefClass.new
